File: src/contest/hackerrank/WOC_28_E.cpp
#include <bits/stdc++.h>

#define mp make_pair
#define pb push_back
#define INF 1<<30
#define MOD 1000000007
#define rint(x) scanf("%d", &(x))
#define rlong(x) scanf("%lld", &(x))
#define SIZE 10000000
#define l(x) x << 1
#define r(x) x << 1|1
#define m(x, y) (x + y)/2

using namespace std;

const int FULL_MASK = 0b1111111111111111111111111111111;

typedef long long ll;
typedef pair<int, int> pi;
typedef pair<ll, ll> pll;

int N, K;
unordered_map<int, double> dp;
string in;
double lookup[SIZE];
int rev (int mask, int len) {
    int ret = 0;
    for (int i = 0; i < len; i++)
        if ((mask & 1 << i) > 0)
            ret |= 1 << (len - i - 1);
    return ret;
}

int remove (int bitmask, int i) {
    int lowerMask = (1 << i) - 1;
    int upperMask = FULL_MASK ^ lowerMask ^ 1 << i;
    return (bitmask & upperMask) >> 1 | (bitmask & lowerMask);
}

double compute (int mask, int k) {
    if (k == K)
        return 0.0;
    if ((mask | 1 << (N - k)) < SIZE) {
        if (lookup[mask | 1 << (N - k)] != -1)
            return lookup[mask | 1 << (N - k)];
    } else {
        if (dp.find(mask | 1 << (N - k)) != dp.end())
            return dp[mask | 1 << (N - k)];
    }

    if ((rev(mask, N - k) | 1 << (N - k)) < SIZE) {
        if (lookup[rev(mask, N - k) | 1 << (N - k)] != -1)
            return lookup[rev(mask, N - k) | 1 << (N - k)];
    } else {
        if (dp.find(rev(mask, N - k) | 1 << (N - k)) != dp.end())
            return dp[rev(mask, N - k) | 1 << (N - k)];
    }

    double ret = 0;

    for (int i = 0; i < (N - k + 1) / 2; i++) {
        bool a = (mask & 1 << i) == 0;
        bool b = (mask & 1 << (N - k - i - 1)) == 0;
        int maskA = remove(mask, i);
        int maskB = remove(mask, N - k - i - 1);
        double prob = ((N - k) % 2 == 1 && i == (N - k) / 2) ? 1.0 / (N - k) : 2.0 / (N - k);

        ret += max(compute(maskA, k + 1) + !a, compute(maskB, k + 1) + !b) * prob;
    }
    if ((mask | 1 << (N - k)) < SIZE)
        lookup[mask | 1 << (N - k)] = ret;
    else
        dp[mask | 1 << (N - k)] = ret;
    return ret;
}

int main () {
    ios_base::sync_with_stdio(false); cin.tie(NULL);
	//freopen("in.txt", "r", stdin);
	//freopen("out.txt", "w", stdout);
    fill_n(lookup, SIZE, -1);
    dp.reserve(2048);
    cin >> N >> K;

    int initial = 0;
    cin >> in;
    for (int i = 0; i < N; i++)
        if (in[i] == 'W')
            initial |= 1 << i;
    printf("%.9f\n", compute(initial, 0));
	return 0;
}

File: src/contest/hackerrank/WOC_28_E.java
package contest.hackerrank;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.util.HashMap;
import java.util.StringTokenizer;

public class WOC_28_E {

  static final int FULL_MASK = Integer.MAX_VALUE;
  static BufferedReader br;
  static PrintWriter out;
  static StringTokenizer st;
  static int N, K;
  static HashMap<Integer, Double> dp = new HashMap<Integer, Double>();

  public static void main(String[] args) throws IOException {
    br = new BufferedReader(new InputStreamReader(System.in));
    out = new PrintWriter(new OutputStreamWriter(System.out));
    //br = new BufferedReader(new FileReader("in.txt"));
    //out = new PrintWriter(new FileWriter("out.txt"));

    N = readInt();
    K = readInt();

    int initial = 0;
    String input = next();

    for (int i = 0; i < N; i++)
      if (input.charAt(i) == 'W')
        initial |= 1 << i;

    out.println(compute(initial, 0));
    out.close();
  }

  static double compute(int mask, int k) {
    if (k == K)
      return 0.0;
    if (dp.containsKey(mask | 1 << (N - k)))
      return dp.get(mask | 1 << (N - k));
    if (dp.containsKey(reverse(mask, N - k) | 1 << (N - k)))
      return dp.get(reverse(mask, N - k) | 1 << (N - k));
    double ret = 0;
    for (int i = 0; i < (N - k + 1) / 2; i++) {
      int a = (mask & 1 << i) > 0 ? 1 : 0;
      int b = (mask & 1 << (N - k - i - 1)) > 0 ? 1 : 0;
      int maskA = remove(mask, i);
      int maskB = remove(mask, N - k - i - 1);
      double prob = ((N - k) % 2 == 1 && i == (N - k) / 2) ? 1.0 / (N - k) : 2.0 / (N - k);

      ret += Math.max(compute(maskA, k + 1) + a, compute(maskB, k + 1) + b) * prob;
    }
    dp.put(mask | 1 << (N - k), ret);
    return ret;
  }

  static int reverse(int mask, int len) {
    int ret = 0;
    for (int i = 0; i < len; i++)
      if ((mask & 1 << i) > 0)
        ret |= 1 << (len - i - 1);
    return ret;
  }

  static int remove(int bitmask, int i) {
    int upperMask = FULL_MASK ^ ((1 << (i + 1)) - 1);
    int lowerMask = (1 << i) - 1;

    return (bitmask & upperMask) >> 1 | (bitmask & lowerMask);
  }

  static String next() throws IOException {
    while (st == null || !st.hasMoreTokens())
      st = new StringTokenizer(br.readLine().trim());
    return st.nextToken();
  }

  static long readLong() throws IOException {
    return Long.parseLong(next());
  }

  static int readInt() throws IOException {
    return Integer.parseInt(next());
  }

  static double readDouble() throws IOException {
    return Double.parseDouble(next());
  }

  static char readCharacter() throws IOException {
    return next().charAt(0);
  }

  static String readLine() throws IOException {
    return br.readLine().trim();
  }
}
